### ISSUE TITLE
Every Tool Review

### ISSUE THREAD

**Issue by Mondaley** [opentoonz/opentoonz#5221]
This post might be a little too late since version v1.8 is almost out, but I am confident that everything discussed here will be covered in v1.9. Without other distractions, let's talk about all tools from the Tool Bar provided by OpenToonz and see where they can be improved. 

_____________________________

### **1.  Animation Tool**
![image](https://github.com/opentoonz/opentoonz/assets/122384675/50473b78-1734-4857-82cc-ebec5bbd876f)

While there are many options available for this tool, the first thing I observe is that there is no grid to snap my drawings. If I want to move something to (0,0) or any other coordinates I'd have to do an approximation with my eyes because we have no grid. I understand that "x" and "y" are shown in the tool option bar, but they are local, not global.

All variables (position, rotation, scale, shear, center) should have snapping features in general.

![image](https://github.com/opentoonz/opentoonz/assets/122384675/7913da67-32eb-4a69-8935-daceb4f66a77)

Also with so many options it would be nice if I could extend the tool option bar downwards so I can see all of them.

![image](https://github.com/opentoonz/opentoonz/assets/122384675/b50281db-a8dd-44df-8105-bcf92c320c4b)

Another thing that would make the Animation Tool more useful would be if you could rotate your 2D sprites and camera in the 3D view to create 3D-like environments. This request would imply updating the 3D view into a real one with the View in Perspective and not in Parallel.  Currently, you can rotate only on the Z-axis. I am not saying that modeling should be implemented (that is tough), I am emphasizing the potential that the 3D view could have.

![image](https://github.com/opentoonz/opentoonz/assets/122384675/3a98c128-5413-458e-9463-ab78f9c0aeab)

### **2. Selection Tool**

![image](https://github.com/opentoonz/opentoonz/assets/122384675/757d9b32-e5d6-4987-b8fd-f14c23f43a65)

Nothing is intriguing to say about it. It is great as it is.

### **3. Brush Tool**

![image](https://github.com/opentoonz/opentoonz/assets/122384675/1a9e81ec-cbf2-4eba-8bb7-c2e581bd6dd6)

The shortcut Ctrl+Alt, which changes the brush size, doesn't work as intended. It wants to change the min and max values at the same time resulting in an unpractical and glitchy feature. Ctrl+Alt should tweak the max value and for the min value, something like Ctrl+Shift should take care of it. Also, this hotkey works only with the Raster Toonz Level but not for the other two.

https://github.com/opentoonz/opentoonz/assets/122384675/35abc287-a727-4fb7-95c1-5996ed89abe3

And if we are still at the "Size" topic, the following shortcuts are slow and unnecessary because Ctrl+alt and Cthr+Shith would do the work:

![image](https://github.com/opentoonz/opentoonz/assets/122384675/3f5824c0-bd72-4d55-898e-89687a63c55d)

Another questionable option that the Brush Tool owns is "Range"., available only on Vector Level.

![image](https://github.com/opentoonz/opentoonz/assets/122384675/e6ec5ebf-6518-4202-b2dc-c2c7d7b47e3e)

Its purpose is to tween between two lines from two different drawings in the Level Strip. The problem is not only the tweening isn't clean (requirements for decent tweening: same number of vertices, same relative position of the vertices, same flow, and the Primary Vertex relative position to be similar otherwise the drawing will flip), but you can't edit the line in any way, making it difficult to work with.

https://github.com/opentoonz/opentoonz/assets/122384675/59b62a09-cedb-42e4-bc05-0100c79608f2

Moreover, why can't custom brushes be added in the Style Editor without messing with the app's files?  It should be directly integrated.

![image](https://github.com/opentoonz/opentoonz/assets/122384675/4bb0d293-aa85-4b8c-a538-9818e6a13d1a)

### **4. Geometric Tool**

![Screenshot 2023-12-15 173834](https://github.com/opentoonz/opentoonz/assets/122384675/8890354a-de58-4289-935a-26abb5570ff8)

Ctrl+Alt is needed for the "Size" and "Thickness".

![Screenshot 2023-12-15 164402](https://github.com/opentoonz/opentoonz/assets/122384675/b778ef95-684c-4c90-870d-627804115e60)
![Screenshot 2023-12-15 164417](https://github.com/opentoonz/opentoonz/assets/122384675/3b028515-d46d-44c9-ac5b-ca2b49fdd06a)

### **5. Type Tool**

![Screenshot 2023-12-15 173951](https://github.com/opentoonz/opentoonz/assets/122384675/0b4b1c74-7b5f-464a-9f6c-669848e73d2a)

The type tool is fine, but it has some limitations. For instance: I can't select the text, move through words with Ctrl+Arrow keys, underline, cut, bold, make it italic, input a custom size, make proper paragraphs, align it to the left, right, middle, and so on...

It is too basic.

For a solid reference of how it should function Microsoft Word is a good candidate.

### **6. Fill Tool**

![Screenshot 2023-12-15 174015](https://github.com/opentoonz/opentoonz/assets/122384675/9a113d09-e5ee-4495-abd2-5d45ac552154)

Wonderful! 🌞

### **7. Paint Brush Tool**

![Screenshot 2023-12-15 174224](https://github.com/opentoonz/opentoonz/assets/122384675/9adb00aa-1498-4d64-85aa-14f8f5d2e340)

At first, I didn't understand its purpose, I thought it was a duplicate of the Brush Tool for Toonz Raster Level only.  It turns out that the Brush Tool only creates strokes that need to be checked by the software every second which is not optimal. On the other side, The Paint Brush Tool, which instead of creating outlines, paints on a single canvas without stroke checking, making the workflow more logical.

This Tool is interesting and worth using.

And, as always, Ctrl+Alt is needed for the "Size".

![image](https://github.com/opentoonz/opentoonz/assets/122384675/639d5eb9-96d6-4ebd-b2b9-15f76813b20c)

### **8. Eraser Tool**

![image](https://github.com/opentoonz/opentoonz/assets/122384675/53cf3a1b-f3b1-430a-947f-92103ccf223d)

I find myself using the Selection Tool and the Eraser option from the Brush Tool more often than the genuine Eraser Tool to delete drawings, but It is intuitive to have, especially for the Toonz Raster Level.

### **9. Tape Tool**

![image](https://github.com/opentoonz/opentoonz/assets/122384675/c714030d-36c7-4e80-a010-04fcc7636741)

If you work in Vector Levels and don't use the Tape Tool then what are you doing?  Taping your vector will make your life so much easier.

One thing that bothers me, though, is the gap check from the view menu.

![image](https://github.com/opentoonz/opentoonz/assets/122384675/6e7fdc4a-e244-45e2-8b2b-d607fb52fa5d)

![image](https://github.com/opentoonz/opentoonz/assets/122384675/d925e4b2-4c44-4806-91c3-ffe2d078c0a7)

When I want to tape in rectangular mode, the result always turns into a mess. Everywhere the software sees a gap tapes it. This feature needs more work.

https://github.com/opentoonz/opentoonz/assets/122384675/da0dcb03-5160-4b5c-8f96-74aff7a496c2

### **10. Finger Tool**

![image](https://github.com/opentoonz/opentoonz/assets/122384675/3cf72007-2b26-48c7-8b43-8c2e78decd43)

This one is pretty much ambiguous, It is only available for Toonz Raster Levels and has no documentation in the OT manual. 
After some research, I have found that it is supposed to fill unfinished areas.

![image](https://github.com/opentoonz/opentoonz/assets/122384675/df4d6f75-c0f5-4923-acbb-d7e2f37e1605)

But if you try to use it yourself, you will realize that it is not even doing its job well. It is a glitchy unfinished tool.

https://github.com/opentoonz/opentoonz/assets/122384675/a82bac1d-1754-4dd1-96fd-eb12964bdfc0

I see its potential, but doesn't seem to be very stable.

### **11. Style Picker Tool**

![image](https://github.com/opentoonz/opentoonz/assets/122384675/8b216f02-d689-421b-8330-b7517d07b89c)

So stylish! ✨

### **12. RGB Picker Tool**

![image](https://github.com/opentoonz/opentoonz/assets/122384675/61e037ba-0499-492f-8e0a-ee4427021564)

If I were to point out one thing, I would like to pick colors from everywhere on my screen not only from the viewport.

### **13. Ruler Tool**

![image](https://github.com/opentoonz/opentoonz/assets/122384675/72a7299b-8a78-47e0-b3e2-2288e5c5efce)

Having the measured variables in the tool options bar and not being able to modify them by hand is exhausting

![image](https://github.com/opentoonz/opentoonz/assets/122384675/24c35502-a6da-4126-b817-8732a02f6602)

Let's have some proper annotations:

![image](https://github.com/opentoonz/opentoonz/assets/122384675/86f9c819-2dd9-4d1e-a09e-019712d59ace)

And also, where is that beautiful grid with its snapping features?

![image](https://github.com/opentoonz/opentoonz/assets/122384675/da004647-5775-4381-b371-f577e7eef713)

### **14. Control Point Editor Tool**

![image](https://github.com/opentoonz/opentoonz/assets/122384675/7c05f503-34a4-461c-8e5c-5d37358cd680)

14.1. Obvious problem there  :)  

![image](https://github.com/opentoonz/opentoonz/assets/122384675/f80f6812-7c61-4089-b763-620ac99f4cc7)
 
It is the Primary Vertex. If you want to read more about it go here:
https://github.com/opentoonz/opentoonz_docs/issues/162

Some consequences of the Primary Vertex:

a. Z-fighting:
https://github.com/opentoonz/opentoonz/assets/122384675/95f2bf71-be21-4c31-83f1-d550f330b56d

b. Data caching:
https://github.com/opentoonz/opentoonz/assets/122384675/3cad6982-3ac8-4f01-a07e-7e157e59b045

c. Vertex rendering number conservation:
https://github.com/opentoonz/opentoonz/assets/122384675/df4e9d7e-ce0c-4245-8aae-ba6baee76e5b

d. Fill-blocking:
https://github.com/opentoonz/opentoonz/assets/122384675/27e90ac4-1d28-4e91-8992-5e06c4542854

And many others...

14.2. I wish I could modify the vertices of multiple Vector Shapes at the same time.

14.3. Every time a Vector Shape crosses over itself new headaches are created. A simple solution to this would be to create a vertex at those intersections. It would be connected to more than two vertices, like in the normal case, but problems will be avoided. Here is a visualization of how it would look like:

https://github.com/opentoonz/opentoonz/assets/122384675/dad278cf-bb67-4849-8c4c-6dec12eb3c17

### **15. Pinch Tool**

![image](https://github.com/opentoonz/opentoonz/assets/122384675/26f8dde2-fc74-41f4-ada5-e748711481f6)

"Corner" and "Size" options need to be consistent. It always confuses me. The fact that for "Corner" to be lesser you need to move the bar's handle to the right while for the "Size" to the left.

![Comparisson](https://github.com/opentoonz/opentoonz/assets/122384675/9f522add-ee56-4ae6-9716-9d5a573587f3)

Ctr+Alt is also needed here for "Corner" and  "Size". Changing the size value by the gizmo provided is slow:

![image](https://github.com/opentoonz/opentoonz/assets/122384675/ebba7673-e8c6-487a-a2ef-0aa6f321d54d)

Lastly, it is hard to pick the end of a line at the intersection with another Vertex Shape/ the ends should be snappy and easy to get hold of:

https://github.com/opentoonz/opentoonz/assets/122384675/9857b96c-4417-4849-bf2c-7f8575e4d41a

### **16. The Pump Tool**

![image](https://github.com/opentoonz/opentoonz/assets/122384675/c340f04a-57e8-4f0f-b35f-e14b332b2bb6)

The only inconvenience, besides Ctrl+Alt for "Size", is that when you enter OT and select the Pump Tool for the first time it won't let you use it unless you change one setting from the Tool Options Bar.

### **17. Magnet Tool**

![image](https://github.com/opentoonz/opentoonz/assets/122384675/e2b48e42-eb9b-450e-8321-833310d58351)

Ctrl+Alt for "Size"

### **18. Bender Tool**

![image](https://github.com/opentoonz/opentoonz/assets/122384675/c0b605e0-edc7-4506-9906-bdf27db03539)

![image](https://github.com/opentoonz/opentoonz/assets/122384675/6bdfd1a1-7228-45e3-a3bb-4b594bdfe307)

https://github.com/opentoonz/opentoonz/assets/122384675/468b0c11-4e42-4fd5-8909-90af9ae124f8

🔥🔥The best tool so far 🔥🔥

### **19. Iron Tool**

![image](https://github.com/opentoonz/opentoonz/assets/122384675/8621e2b7-9489-4ffd-8cee-6148985aec2d)

No doubt, really handy! 

### **20. CutterTool**

![image](https://github.com/opentoonz/opentoonz/assets/122384675/d363e4f3-4707-41c3-9402-f32d7e9e5104)

Can't live without it!

### **21 Skeleton Tool**

![image](https://github.com/opentoonz/opentoonz/assets/122384675/ce99ef4c-6f78-41c1-8925-b3faea2b3819)

This tool on its own can make basic riggs only. The resulting bones have no constraints or other fancy options.

https://github.com/opentoonz/opentoonz/assets/122384675/1e9531f7-fe42-440c-88ab-f9548b8d5824

I don't use it since it takes so many Levels to make one character. I know that I can collapse those Levels into one, but imagine how messy your scene will become with so many Levels in the X-sheet or Timeline.

Another thing to consider is that the connection in the Schematic Stage has options that I believe to be unnecessary:

https://github.com/opentoonz/opentoonz/assets/122384675/4704dfac-da71-4f01-859b-c7a53cf90646

The only reasonable letter to use is "B", which stands for "Bone". The other ones: "A", "C", "D", ...., "Z" make no sense. They just move the Level along the X-axis. To whom was this feature useful?

And bonus crash: zoom all the way into a node from the schematic stage.

https://github.com/opentoonz/opentoonz/assets/122384675/4a08ca9c-fe63-485f-a56c-5222e6a4a613

It was the crash pop-up window there but my recording app didn't show it.

I would prefer to have a whole skeleton for a single normal Vector Level rather than a collapsed Level with tens of other Levels.

### **22 Tracker Tool**

![image](https://github.com/opentoonz/opentoonz/assets/122384675/a796350a-06e9-4062-9ba4-71b23f5f78ee)

Haven't used this one too frequently, but I understand its role: it is a hook that moves along a specified shape within an image. This tool is as good as the algorithm. From what I know it is doing its job but your video must be slow for more accuracy.

### **23 The Hook Tool**

![image](https://github.com/opentoonz/opentoonz/assets/122384675/0864d3a1-36d5-47c4-976b-0a99784bd9f2)

Hooks are just saved positions where you can link your Levels. In the Stage Schematic, they are referred to by numbers instead of letters. If you think about it every node link option from the Stage Schematic can be called a hook. For example: you can say that the "bone" ("B") is also a hook but it can be connected with other hooks and be influenced by them.

### **24 Plastic Tool**

![image](https://github.com/opentoonz/opentoonz/assets/122384675/d8f279d1-b53c-4160-a071-cfab20b8b823)

This is by far one of the tools with the most potential. Having flexible drawings, in many cases, is truly time-saving. Although, as much as I enjoyed playing with it, I couldn't observe the lack of features it has. I believe this should be the ultimate tool when it comes to rigging.

The workflow with the Plastic Tool shows how a genuine model should be constructed. Needing one level instead of many others is the way to go.

![image](https://github.com/opentoonz/opentoonz/assets/122384675/845085df-5bfc-44eb-b683-b2498994b84f)

Spreading your drawing before using the Plastic Tool on it should be your first rule of thumb. You don't want the mesh to fuse in unwanted places. 

![image](https://github.com/opentoonz/opentoonz/assets/122384675/f926c652-acdb-461e-acbd-c44690621f08)

At the moment we have triangulated mesh, but maybe we will get a quad-based mesh in the future.

Looking further, there is no mirror option for building bones, sadly.

![image](https://github.com/opentoonz/opentoonz/assets/122384675/24702d2b-407d-48c7-b24e-46a077418f85)

And we have no inverse kinematics like the bone tool does:

![image](https://github.com/opentoonz/opentoonz/assets/122384675/7059b0de-77ca-4cfd-a86c-8760ac3297fb)

One thing I have learned, besides many others by using Blender, is that the typology matters for any mesh.  If your character/object is not moving/bending as intended, it might be because of the poorly made mesh.

![tipology](https://github.com/opentoonz/opentoonz/assets/122384675/80e46b05-7438-4a0d-ba9e-cbfd437c7937)

As you can observe, the quad-like mesh is better since it is easier to manipulate and make decisions on how your subject bends.

Speaking of topology, the current workflow of modifying meshes is not ideal: every action must be done by right-clicking and then selecting your action with a left-click. In addition, you can select one vertex/edge at a time. I would want similar flexibility as Blender offers: selecting multiple edges, vertices, faces, invert selection, extruding, deleting whatever you want, inserting whatever you want wherever you want, and so on...

![image](https://github.com/opentoonz/opentoonz/assets/122384675/4131f7e0-0cc0-46fd-9e28-45d805db8379)

For the Paint Rigid mode, I think adding values between 0 and 1 won't hurt anyone.

Moving on, I like the "SO" option. It is the Sorting Order and can be adjusted for every vertex. I wish the SO value was stored within every vertex after I copy the skeleton and paste it somewhere else.

If you have multiple drawings of the same character, you cannot copy and paste your mesh across them. You can either use the same frame or generate other meshes, which adds more manual tweaking.

![props](https://github.com/opentoonz/opentoonz/assets/122384675/9b3b69d7-a16a-4a9e-a959-577d424004b9)

The drawings with a Christmas hat and a cape would need to be made in separate layers and have some physics applied to them. But that is another story that needs to be taken care of by the 2D 360 Character rigging: https://github.com/opentoonz/opentoonz_docs/issues/173

### **25 Zoom Tool**

![image](https://github.com/opentoonz/opentoonz/assets/122384675/ee8077b0-3892-4d8b-b844-a9540f68508b)

Perfect!⭐

### **26 Rotate Tool**

Needs some snapping every 15° or an angle given by the user.

### **28 Rotate Tool**

![image](https://github.com/opentoonz/opentoonz/assets/122384675/fcfb366f-426f-4148-807f-20673691df65)

Gorgeous! 🌈

_____________________________

And I think that's it! My view on all of them. If I missed something let us know in the comments.
OT is still in development and needs every single feedback from the community. I hope this post will help move forward this software to better versions.

**Comment by RodneyBaker** [comment]
The release of v1.8 is still some time away so now is the perfect time to plunder the depths of all things Opentoonz.
Thanks for adding your voice to the conversation.

**Comment by RodneyBaker** [comment]
Just did a quick scrub of this report and this stood out:

> RGB Picker Toolimage
> If I were to point out one thing, I would like to pick colors from everywhere on my screen not only from the viewport.

It's a little hidden but way way over to the right there is a button 'Pick Screen'.
'Pick Screen' allows pixel sampling from anywhere on the screen.

**Comment by RodneyBaker** [comment]
All good. It's great feedback and insight from someone that was indeed learning the program and sharing that experience.  That's very valuable information to have.

Everyone's experience is a little different.  Thanks for sharing yours!
